server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -290,18 +300,28 @@ public DataStream replaceBackingIndex(Index existingBackingIndex, Index newBacki
                 String.format(Locale.ROOT, "index [%s] is not part of data stream [%s]", existingBackingIndex.getName(), name)
             );
         }
-        if (generation == (backingIndexPosition + 1)) {
+        if (indices.size() == (backingIndexPosition + 1)) {
```

Comment:
@martijnvg, when you have a minute, could you review this one? The main change is in this line though there are a couple others that I mentioned in the PR description.